crates/vm/levm/src/opcode_handlers/environment.rs

```diff
@@ -230,29 +230,52 @@ impl VM {
             .try_into()
             .map_err(|_| VMError::VeryLargeNumber)?;
 
-        let gas_cost =
-            gas_cost::codecopy(current_call_frame, size, dest_offset).map_err(VMError::OutOfGas)?;
+        let gas_cost = gas_cost::codecopy(current_call_frame, size, destination_offset)
+            .map_err(VMError::OutOfGas)?;
 
         self.increase_consumed_gas(current_call_frame, gas_cost)?;
 
         if size == 0 {
             return Ok(OpcodeSuccess::Continue);
         }
 
-        let bytecode_len = current_call_frame.bytecode.len();
-        let code = if offset < bytecode_len {
-            current_call_frame.bytecode.slice(
-                offset
-                    ..(offset.checked_add(size).ok_or(VMError::Internal(
+        let new_memory_size = (destination_offset
+            .checked_add(size)
+            .ok_or(VMError::Internal(
+                InternalError::ArithmeticOperationOverflow,
+            ))?)
+        .checked_next_multiple_of(WORD_SIZE)
+        .ok_or(VMError::Internal(
+            InternalError::ArithmeticOperationOverflow,
+        ))?;
+        let current_memory_size = current_call_frame.memory.data.len();
+
+        if current_memory_size < new_memory_size {
+            current_call_frame
+                .memory
+                .data
+                .try_reserve(new_memory_size)
+                .map_err(|_err| VMError::MemorySizeOverflow)?;
+            current_call_frame.memory.data.resize(new_memory_size, 0);
+        }
```

Comment:
What do you think about this replacement:
```suggestion
        current_call_frame.memory.resize(
            destination_offset
                .checked_add(size)
                .ok_or(VMError::VeryLargeNumber)?,
        )?;
```
We have to make Memory's `resize` function public though.

If I'm not mistaken it looks like it has the same behavior